optimizer/src/main/java/com/qubole/quark/planner/MaterializedViewFilterScanRule.java

```diff
@@ -0,0 +1,92 @@
+ * See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package com.qubole.quark.planner;
+
+import org.apache.calcite.plan.MaterializedViewSubstitutionVisitor;
+import org.apache.calcite.plan.RelOptMaterialization;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptRuleOperand;
+import org.apache.calcite.plan.hep.HepPlanner;
+import org.apache.calcite.plan.hep.HepProgram;
+import org.apache.calcite.plan.hep.HepProgramBuilder;
+import org.apache.calcite.plan.volcano.VolcanoPlanner;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.rel.rules.FilterProjectTransposeRule;
+import org.apache.calcite.rel.rules.ProjectMergeRule;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Planner rule that converts
+ * a {@link org.apache.calcite.rel.core.Filter}
+ * on a {@link org.apache.calcite.rel.core.TableScan}
+ * to a filter on Materialized View
+ */
+public abstract class MaterializedViewFilterScanRule extends RelOptRule {
+  /** Rule that matches Filter on TableScan. */
+  public static final MaterializedViewFilterScanRule INSTANCE =
+      new MaterializedViewFilterScanRule(
+          operand(Filter.class,
+              operand(TableScan.class, null, none())),
+          "MaterializedViewFilterTableRule") {
+        public void onMatch(RelOptRuleCall call) {
+          final Filter filter = call.rel(0);
+          final TableScan scan = call.rel(1);
+          apply(call, filter, scan);
+        }
+      };
+  private final HepProgram program = new HepProgramBuilder()
+      .addRuleInstance(FilterProjectTransposeRule.INSTANCE)
+      .addRuleInstance(ProjectMergeRule.INSTANCE)
+      .build();
+
+  //~ Constructors -----------------------------------------------------------
+
+  /** Creates a FilterTableRule. */
+  protected MaterializedViewFilterScanRule(RelOptRuleOperand operand, String description) {
+    super(operand, description);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  protected void apply(RelOptRuleCall call, Filter filter, TableScan scan) {
+    RelNode root = filter.copy(filter.getTraitSet(),
+        Collections.singletonList((RelNode) scan));
+    RelOptPlanner planner = call.getPlanner();
+    if (planner instanceof VolcanoPlanner) {
+      List<RelOptMaterialization> materializations
+          = ((VolcanoPlanner) planner).getApplicableMaterializations();
+      for (RelOptMaterialization materialization : materializations) {
+        RelNode target = materialization.queryRel;
+        final HepPlanner hepPlanner =
+            new HepPlanner(program, planner.getContext());
+        hepPlanner.setRoot(target);
+        target = hepPlanner.findBestExp();
```

Comment:
For my information, this will return the original plan or the optimized plan with MV ?